docs/proposals/01-firmware-update.md

```diff
@@ -0,0 +1,381 @@
+    s4 --> [*]
+    s5 --> [*]
+```
+
+### Update service
+
+This is an application providing an API to schedule, execute and collect results of firmware update, discover and scan jobs.
+It consists of the following components:
+
+- [API server](#api-server)
+- [scheduler](#scheduler)
+- [job runner](#job-runner)
+- [update/scan/discovery jobs](#updatescandiscovery-job) (concrete implementations)
+
+Update service MUST be extensible in part of the possibility of using plugins for update strategy.
+At the same time, the only update strategy can be enabled for a particular update service instance.
+Update service MAY also include the component to download and store the firmwares.
+
+#### API server
+
+API server exposes update service endpoints and forwards incoming requests to the scheduler.
+It MUST expose the following endpoints:
+
+- Update(UpdateRequest) UpdateResponse;
+  - `UpdateRequest` MUST contain the reference to concrete `Server` object and the list of the firmware-version to be installed.
+  - `UpdateResponse` MUST contain the status of the request with error code if any.
+- Scan(ScanRequest) ScanResponse;
+  - `ScanRequest` MUST contain the reference to concrete `Server` object.
+  - `ScanResponse` MUST contain the status of the request with error code if any.
+- Discover(DiscoverRequest) DiscoverResponse;
+  - `DiscoverRequest` MUST contain the reference to concrete `DiscoveredFirmware` object.
+  - `DiscoverResponse` MUST contain the status of the request with error code if any.
+- UpdateServer(UpdateServerRequest) UpdateServerResponse;
+  - `UpdateServerRequest` MUST contain the reference to concrete `Server` object and the list of installed firmware-versions.
+  This endpoint MUST be used by update or scan jobs after a task if finished to send results and invoke the object's update.
+  - `UpdateServerResponse` MUST contain the status of the request with error code if any.
+- UpdateDiscoveredFirmware(UpdateDiscoveredFirmwareRequest) UpdateDiscoveredFirmwareResponse;
+  - `UpdateDiscoveredFirmwareRequest` MUST contain the reference to concrete `DiscoveredFirmware` object and the list of discovered firmware-versions.
+  This endpoint MUST be used by discovery jobs after a task is finished to send results and invoke the object's update.
+  - `UpdateDiscoveredFirmwareResponse` MUST contain the status of the request with error code if any.
+
+Depending on the type of the request, it SHOULD be forwarded to the corresponding scheduler's queue.
+
+#### Scheduler
+
+Scheduler is a component of the update service that is responsible for scheduling jobs:
+
+- it MUST NOT allow running several jobs on the same target server simultaneously;
+- it MAY discard incoming update or scan requests if the same jobs targeting the same server are already scheduled;
+- it MAY discard incoming discovery requests if the job targeting the same manufacturer-model pair is already scheduled;
+- it MUST have a mechanism to limit the number of parallel jobs;
```

Comment:
Did you consider how this would work together with https://github.com/ironcore-dev/metal-operator/issues/76 ? Like, if we schedule a Firmware Upgrade how do we signal this intent to the Workload or Management cluster? 

Related to above, we should get some kind of health status from Workload cluster when performing an update on a set of servers one by one. Like, when first server upgrade is finished we ensure that cluster is `healthy` before proceeding with next server.